dvc/tree/webdav.py

```diff
@@ -0,0 +1,190 @@
+import logging
+import threading
+
+from funcy import cached_property, wrap_prop
+
+from dvc.exceptions import DvcException, WebDAVConfigError
+from dvc.path_info import WebDAVURLInfo
+from dvc.scheme import Schemes
+
+from .base import BaseTree
+from .http import ask_password
+
+logger = logging.getLogger(__name__)
+
+
+class WebDAVTree(BaseTree):  # pylint:disable=abstract-method
+    # Use webdav scheme
+    scheme = Schemes.WEBDAV
+
+    # URLInfo for Webdav ~ replaces webdav -> http
+    PATH_CLS = WebDAVURLInfo
+
+    # Non traversable as walk_files is not implemented
+    CAN_TRAVERSE = False
+
+    # Implementation based on webdav3.client
+    REQUIRES = {"webdavclient3": "webdav3.client"}
+
+    # Constructor
+    def __init__(self, repo, config):
+        # Call BaseTree constructor
+        super().__init__(repo, config)
+
+        # Get username from configuration
+        self.user = config.get("user", None)
+
+        # Get password from configuration (might be None ~ not set)
+        self.password = config.get("password", None)
+
+        # Whether to ask for password is it is not set
+        self.ask_password = config.get("ask_password", False)
+
+        # Webdav root directory
+        self.root = config.get("root", "/")
+
+        # Use token for webdav auth
+        self.token = config.get("token", None)
+
+        # Path to certificate
+        self.cert_path = config.get("cert_path", None)
+
+        # Path to private key
+        self.key_path = config.get("key_path", None)
+
+        # Connection timeout
+        self.timeout = config.get("timeout", 30)
+
+        # Get URL from configuration
+        self.url = config.get("url", None)
+
+        # If URL in config parse path_info
+        if self.url:
+            self.path_info = self.PATH_CLS(self.url)
+
+            # If username not specified try to use from URL
+            if self.user is None and self.path_info.user is not None:
+                self.user = self.path_info.user
+
+            # If username specified add to path_info
+            if self.user is not None:
+                self.path_info.user = self.user
+        else:
+            self.path_info = None
+
+    # Webdav client
+    @wrap_prop(threading.Lock())
+    @cached_property
+    def _client(self):
+        # Import the webdav client library
+        from webdav3.client import Client
+
+        # Construct hostname from path_info
+        hostname = (
+            self.path_info.scheme.replace("webdav", "http")
+            + "://"
+            + self.path_info.host
+        )
```

Comment:
`port` is not passed, try something like following, perhaps?
```python
http_info = HTTPURLInfo(self.path_info.url)
hostname = http_info.replace(path="").url
```

Reply:
Yes, you are right, I somehow forgot that there is more in the `path_info` than `scheme` and `host`... Should be fixed now.